File: mox/src/main/java/com/example/mox/ui/SettingView.java
package com.example.mox.ui;

import android.app.DownloadManager;
import android.content.Context;
import android.net.Uri;
import android.os.Build;
import android.os.FileUtils;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;

import com.example.mox.ManholeConstants;
import com.example.mox.Mox;
import com.example.mox.R;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class SettingView extends LinearLayout implements View.OnClickListener {

    private EditText etPath;
    private Button btnEdit;
    private Button btnRefresh;

    //TODO 2019-11-21 by WangChao
    private String lastPath = "";
    private Context context;

    public SettingView(Context context) {
        super(context);
        init(context);
    }

    public SettingView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public SettingView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public SettingView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(context);
    }

    private void init(Context context) {
        this.context = context;
        View view = LayoutInflater.from(context).inflate(R.layout.layout_setting, this);
        etPath = view.findViewById(R.id.etPath);
        btnEdit = view.findViewById(R.id.btnEdit);
        btnRefresh = view.findViewById(R.id.btnRefresh);

        btnEdit.setOnClickListener(this);
        btnRefresh.setOnClickListener(this);


        etPath.setText(Mox.getInstance().sp.getString(ManholeConstants.KEY_DB_PATH));

    }

    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.btnEdit) {

            if (etPath.isEnabled()) {
                String trim = etPath.getText().toString().trim();
                if (!lastPath.equals(trim)) {
                    //TODO 2019-11-21 by WangChao 保存
                    refresh();
                }
                etPath.setEnabled(false);
                btnEdit.setText("编辑");
            } else {
                etPath.setEnabled(true);
                etPath.setSelection(etPath.getText().length());
                btnEdit.setText("保存");
            }

        } else if (v.getId() == R.id.btnRefresh) {
            refresh();
        }
    }

    private void refresh() {
        String trim = etPath.getText().toString().trim();
        if (trim.startsWith("http")) {
            downloadDbFile(trim);
            Mox.getInstance().sp.put(ManholeConstants.KEY_DB_PATH, trim);
        } else if (trim.startsWith("/")) {
            copyDBFile(trim);
            Mox.getInstance().initDb(context, context.getFilesDir().getAbsolutePath() + File.separator + ManholeConstants.DB_NAME);
            Mox.getInstance().sp.put(ManholeConstants.KEY_DB_PATH, trim);
        }
    }

    private void  downloadDbFile(String url) {
        final Request request = new Request.Builder().url(url).header("Cache-Control","no-cache").get().build();
        OkHttpClient client = new OkHttpClient.Builder().build();
        client.newCall(request).enqueue(new Callback() {

            @Override
            public void onFailure(Call call, final IOException e) {
                post(new Runnable() {
                    @Override
                    public void run() {
                        Toast.makeText(context,"fail:" + e.getClass().getSimpleName(), Toast.LENGTH_SHORT).show();
                    }
                });
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                if (response != null) {
                    int code = response.code();
                    if (code != 200 || response.body() == null) {
                        return;
                    }

                    InputStream inputStream = response.body().byteStream();
                    File dbFile = new File(context.getFilesDir(), ManholeConstants.DB_NAME);
                    if (!dbFile.exists()) {
                        try {
                            dbFile.createNewFile();
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                    }

                    OutputStream outputStream = null;
                    try {
                        outputStream = new FileOutputStream(dbFile);
                        byte[] buffer = new byte[4096];
                        int len = -1;
                        while ((len = inputStream.read(buffer)) > 0) {
                            outputStream.write(buffer,0,len);
                        }
                        Mox.getInstance().initDb(context, context.getFilesDir().getAbsolutePath() + File.separator + ManholeConstants.DB_NAME);
                        post(new Runnable() {
                            @Override
                            public void run() {
                                Toast.makeText(context, "complete",Toast.LENGTH_SHORT).show();
                            }
                        });
                    } catch (FileNotFoundException e) {
                        e.printStackTrace();
                    } catch (IOException e) {
                        e.printStackTrace();
                    } finally {
                        if (inputStream != null) {
                            try {
                                inputStream.close();
                            }catch (IOException e) {

                            }
                        }
                        if (outputStream != null) {
                            try {
                                outputStream.close();
                            }catch (IOException e) {

                            }
                        }
                    }

                }
            }
        });

    }

    private void copyDBFile(String path) {
        File file = new File(path);
        if (file.exists()) {

            File dbFile = new File(context.getFilesDir(), ManholeConstants.DB_NAME);
            if (!dbFile.exists()) {
                try {
                    dbFile.createNewFile();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }

            InputStream inputStream = null;
            OutputStream outputStream = null;
            try {
                inputStream = new FileInputStream(file);
                outputStream = new FileOutputStream(dbFile);
                FileUtils.copy(inputStream, outputStream);
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            } catch (IOException e) {
                e.printStackTrace();
            } finally {
                if (inputStream != null) {
                    try {
                        inputStream.close();
                    }catch (IOException e) {

                    }
                }
                if (outputStream != null) {
                    try {
                        outputStream.close();
                    }catch (IOException e) {

                    }
                }
            }

        }
    }
}

File: manhole/src/main/java/com/freesith/manhole/ui/SettingView.kt
package com.freesith.manhole.ui

import android.content.Context
import android.os.Build
import android.os.FileUtils
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout
import android.widget.Toast
import androidx.annotation.RequiresApi
import com.freesith.manhole.ManholeMock
import com.freesith.manhole.ManholeConstants
import com.freesith.manhole.R
import com.freesith.manhole.ext.default
import com.freesith.manhole.history.ManholeHistory
import com.freesith.manhole.util.ManholeSp
import kotlinx.android.synthetic.main.layout_setting.view.*
import okhttp3.*
import java.io.*

class SettingView : LinearLayout, View.OnClickListener {

    //TODO 2019-11-21 by WangChao
    private val lastPath = ""
    private var thisContext: Context? = null

    constructor(context: Context) : super(context) {
        init(context)
    }

    constructor(
        context: Context,
        attrs: AttributeSet?
    ) : super(context, attrs) {
        init(context)
    }

    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int
    ) : super(context, attrs, defStyleAttr) {
        init(context)
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int
    ) : super(context, attrs, defStyleAttr, defStyleRes) {
        init(context)
    }

    private fun init(context: Context) {
        this.thisContext = context
        val view = LayoutInflater.from(context).inflate(R.layout.layout_setting, this)
        manhole_btnEdit?.setOnClickListener(this)
        manhole_btnRefresh?.setOnClickListener(this)
        manhole_etPath?.setText(ManholeSp.dbPath.default("http://apk.minim.red/moc.db"))
        manhole_btnEdit?.setVisibility(View.GONE)

        manhole_swHistory.isChecked = ManholeSp.enableHistory
        manhole_swHistory.setOnCheckedChangeListener { buttonView, isChecked ->
            ManholeSp.enableHistory = isChecked
        }
        manhole_swSummon.isChecked = ManholeSp.enableSummon
        manhole_swSummon.setOnCheckedChangeListener { buttonView, isChecked ->
            ManholeSp.enableSummon = isChecked
        }
        manhole_swShortcut.isChecked = ManholeSp.enableHistoryShortcut
        manhole_swShortcut.setOnCheckedChangeListener { buttonView, isChecked ->
            ManholeSp.enableHistoryShortcut = isChecked
        }
        manhole_btnClear?.setOnClickListener(this)
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.manhole_btnEdit -> {
                if (manhole_etPath!!.isEnabled) {
                    val trim = manhole_etPath!!.text.toString().trim { it <= ' ' }
                    if (lastPath != trim) {
                        //TODO 2019-11-21 by WangChao 保存
                        refresh()
                    }
                    manhole_etPath!!.isEnabled = false
                    manhole_btnEdit!!.text = "编辑"
                } else {
                    manhole_etPath!!.isEnabled = true
                    manhole_etPath!!.setSelection(manhole_etPath!!.text.length)
                    manhole_btnEdit!!.text = "保存"
                }
            }
            R.id.manhole_btnRefresh -> {
                refresh()
            }

            R.id.manhole_btnClear -> {
                ManholeHistory.clearDb()
                Toast.makeText(thisContext, "已清空", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun refresh() {
        val trim = manhole_etPath!!.text.toString().trim { it <= ' ' }
        if (trim.startsWith("http")) {
            downloadDbFile(trim)
            ManholeSp.dbPath = trim
        } else if (trim.startsWith("/")) {
            copyDBFile(trim)
            ManholeMock.initMockDb(
                thisContext,
                thisContext!!.filesDir
                    .absolutePath + File.separator + ManholeConstants.MOCK_DB_NAME
            )
            ManholeSp.dbPath = trim
        }
    }

    private fun downloadDbFile(url: String) {
        val request =
            Request.Builder().url(url).header("Cache-Control", "no-cache").get().build()
        val client = OkHttpClient.Builder().build()
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                post(Runnable {
                    Toast.makeText(
                        thisContext,
                        "fail:" + e.javaClass.simpleName,
                        Toast.LENGTH_SHORT
                    ).show()
                })
            }

            @Throws(IOException::class)
            override fun onResponse(
                call: Call,
                response: Response
            ) {
                if (response != null) {
                    val code = response.code()
                    if (code != 200 || response.body() == null) {
                        return
                    }
                    val inputStream = response.body()!!.byteStream()
                    val dbFile =
                        File(thisContext!!.filesDir, ManholeConstants.MOCK_DB_NAME)
                    if (!dbFile.exists()) {
                        try {
                            dbFile.createNewFile()
                        } catch (e: IOException) {
                            e.printStackTrace()
                        }
                    }
                    var outputStream: OutputStream? = null
                    try {
                        outputStream = FileOutputStream(dbFile)
                        val buffer = ByteArray(4096)
                        var len = -1
                        while ((inputStream!!.read(buffer).also { len = it }) > 0) {
                            outputStream.write(buffer, 0, len)
                        }
                        ManholeMock.initMockDb(
                            thisContext,
                            thisContext!!.filesDir
                                .absolutePath + File.separator + ManholeConstants.MOCK_DB_NAME
                        )
                        post(object : Runnable {
                            override fun run() {
                                Toast.makeText(thisContext, "complete", Toast.LENGTH_SHORT).show()
                            }
                        })
                    } catch (e: FileNotFoundException) {
                        e.printStackTrace()
                    } catch (e: IOException) {
                        e.printStackTrace()
                    } finally {
                        if (inputStream != null) {
                            try {
                                inputStream.close()
                            } catch (e: IOException) {
                            }
                        }
                        if (outputStream != null) {
                            try {
                                outputStream.close()
                            } catch (e: IOException) {
                            }
                        }
                    }
                }
            }
        })
    }

    private fun copyDBFile(path: String) {
        val file = File(path)
        if (file.exists()) {
            val dbFile = File(thisContext!!.filesDir, ManholeConstants.MOCK_DB_NAME)
            if (!dbFile.exists()) {
                try {
                    dbFile.createNewFile()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
            var inputStream: InputStream? = null
            var outputStream: OutputStream? = null
            try {
                inputStream = FileInputStream(file)
                outputStream = FileOutputStream(dbFile)
                FileUtils.copy(inputStream, outputStream)
            } catch (e: FileNotFoundException) {
                e.printStackTrace()
            } catch (e: IOException) {
                e.printStackTrace()
            } finally {
                if (inputStream != null) {
                    try {
                        inputStream.close()
                    } catch (e: IOException) {
                    }
                }
                if (outputStream != null) {
                    try {
                        outputStream.close()
                    } catch (e: IOException) {
                    }
                }
            }
        }
    }
}